File: SwiGi/SwiGi/HIDTransport.swift
import Foundation
import CHIDAPI

enum TransportError: LocalizedError {
    case closed
    case readFailed(String)
    case writeFailed(String)

    var errorDescription: String? {
        switch self {
        case .closed:
            return "Transport is closed"
        case .readFailed(let message):
            return "HID read failed: \(message)"
        case .writeFailed(let message):
            return "HID write failed: \(message)"
        }
    }
}

final class HIDTransport {
    let path: String
    let pid: UInt16
    private var device: OpaquePointer?

    init(path: String, pid: UInt16) throws {
        self.path = path
        self.pid = pid
        guard let dev = path.withCString({ hid_open_path($0) }) else {
            throw TransportError.readFailed(HIDAPIBridge.lastError())
        }
        device = dev
    }

    func read(timeout: Int = 500) throws -> Data? {
        guard let device else { throw TransportError.closed }
        var buffer = [UInt8](repeating: 0, count: HIDPPConstants.maxReadSize)
        let count = buffer.withUnsafeMutableBufferPointer { ptr in
            hid_read_timeout(device, ptr.baseAddress, HIDPPConstants.maxReadSize, Int32(timeout))
        }
        if count < 0 {
            let err = HIDAPIBridge.lastError(device: device)
            if err.lowercased().contains("success") || err.isEmpty {
                return nil
            }
            throw TransportError.readFailed(err)
        }
        guard count > 0 else { return nil }
        return Data(buffer.prefix(Int(count)))
    }

    func write(_ message: Data) throws {
        guard let device else { throw TransportError.closed }
        let count = message.withUnsafeBytes { ptr in
            hid_write(device, ptr.baseAddress?.assumingMemoryBound(to: UInt8.self), message.count)
        }
        if count < 0 {
            throw TransportError.writeFailed(HIDAPIBridge.lastError(device: device))
        }
    }

    func close() {
        if let device {
            hid_close(device)
            self.device = nil
        }
    }

    deinit {
        close()
    }

    var isOpen: Bool { device != nil }
}

File: SwiGi.Win/SwiGi.Win/Hid/HidTransport.cs
namespace SwiGi.Win.Hid;

internal sealed class HidTransport : IDisposable
{
    private IntPtr _device;
    public string Path { get; }
    public ushort Pid { get; }

    public HidTransport(string path, ushort pid)
    {
        Path = path;
        Pid = pid;
        _device = HidApiNative.hid_open_path(path);
        if (_device == IntPtr.Zero)
            throw new IOException($"hid_open_path failed: {HidApiNative.GetLastError()}");
    }

    public bool IsOpen => _device != IntPtr.Zero;

    public byte[]? Read(int timeoutMs = 500)
    {
        if (_device == IntPtr.Zero)
            throw new InvalidOperationException("Transport is closed");

        var buffer = new byte[HidPPConstants.MaxReadSize];
        var count = HidApiNative.hid_read_timeout(_device, buffer, (nuint)buffer.Length, timeoutMs);
        if (count < 0)
        {
            var err = HidApiNative.GetLastError(_device);
            if (err.Contains("success", StringComparison.OrdinalIgnoreCase) || string.IsNullOrEmpty(err))
                return null;
            throw new IOException($"HID read failed: {err}");
        }

        if (count == 0) return null;
        var result = new byte[count];
        Array.Copy(buffer, result, count);
        return result;
    }

    public void Write(ReadOnlySpan<byte> message)
    {
        if (_device == IntPtr.Zero)
            throw new InvalidOperationException("Transport is closed");

        var arr = message.ToArray();
        var count = HidApiNative.hid_write(_device, arr, (nuint)arr.Length);
        if (count < 0)
            throw new IOException($"HID write failed: {HidApiNative.GetLastError(_device)}");
    }

    public void Dispose()
    {
        if (_device != IntPtr.Zero)
        {
            HidApiNative.hid_close(_device);
            _device = IntPtr.Zero;
        }
    }
}
